src/main/java/jssc/SerialPortStream.java

```diff
@@ -0,0 +1,221 @@
+ *
+ * jSSC is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with jSSC.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If you use jSSC in public project you can inform me about this by e-mail,
+ * of course if you want it.
+ *
+ * e-mail: scream3r.org@gmail.com
+ * web-site: http://scream3r.org | http://code.google.com/p/java-simple-serial-connector/
+ *
+ * SerialPortStream.java is written and contributed under the same license
+ * to the jSSC project by FactorIT B.V. in 2022 (factoritbv on github)
+ *
+ */
+package jssc;
+
+import java.net.*;
+import java.util.*;
+import java.util.concurrent.*;
+import java.io.*;
+
+public class SerialPortStream {
+    
+    private SerialPort m_serial_port;
+    private SerialPortInputStream m_serial_port_input_stream;
+    private SerialPortOutputStream m_serial_port_output_stream;
+    private volatile boolean m_closed;
+
+    private class SerialPortInputStream extends InputStream {
+        private byte[] m_buffer;
+        private int m_buffer_pos = 0;
+        private int m_buffer_len = 0;
+
+        public SerialPortInputStream() {
+            m_buffer_len = 0;
+            m_buffer_pos = 0;
+        }
+
+        public void close() throws IOException {
+            SerialPortInputStream.this.close();
+        }
+
+        public int read() throws IOException {
+            try {
+                // See if we run out of available bytes, and try to re-fill the buffer
+                if (m_buffer_pos >= m_buffer_len) {
+                    m_buffer_len = m_serial_port.getInputBufferBytesCount();
+                    if (m_buffer_len == 0) {
+                        // Nothing available, just block until the first byte comes available and return directly
+                        return (int)m_serial_port.readBytes(1)[0];
+                    }
+                    // Fetch the available bytes at once
+                    m_buffer = m_serial_port.readBytes(m_buffer_len);
+                    // Reset the position in the buffer
+                    m_buffer_pos = 0;
+                }
+                return m_buffer[m_buffer_pos++];
+            } catch (SerialPortException ex) {
+                throw new IOException(ex);
+            }
+        }
+
+        public int available() throws IOException {
+            try {
+                return m_serial_port.getInputBufferBytesCount();
+            } catch (SerialPortException ex) {
+                throw new IOException(ex);
+            }
+        }
+    }
+
+    private class SerialPortOutputStream extends OutputStream {
+        
+        public void close() throws IOException {
+            SerialPortStream.this.close();
+        }
```

Comment:
This actually makes sense, because as with the Socket class you will perform call `.getInputStream()` somewhere in the application code and feed that to a another `Stream` or `Reader`. When that is closed, the InputStream will be closed, but that should trigger a complete close (also the OutputStream and the SerialPort) self.

A sanity check in the main `close()` function can be added to check if the Input/Output stream is closed already, but actually the Java docs specify it does not harm to close an Input/Output stream more than once (it is and stays closed).